build/configure/app-configure-step.js

```diff
@@ -0,0 +1,138 @@
+  this.name = this.webapp.sourceDirectoryName;
+  this.options = options;
+  this.buildConfig = new BuildConfig();
+};
+
+AppConfigureStep.prototype = {
+  /**
+   * Configure start from here.
+   */
+  start: function() {
+    var options = this.options;
+    this._preSteps = [];
+    this.appDirFile = utils.getFile(this.appDir);
+    this.xpcshell = utils.getEnv('XPCSHELLSDK');
+
+    // we'll execute our current modules with stringified options.
+    this.appOptions = utils.cloneJSON(options);
+
+    this.appOptions.APP_DIR = this.appDir,
+    this.appOptions.STAGE_APP_DIR = utils.joinPath(this.options.STAGE_DIR,
+      this.appDirFile.leafName);
+    this.appOptions.webapp = this.webapp;
+    // We may encounter error while stringify manifest , put it to command line
+    // and parse it back to json object, so we simply delete these since the
+    // following steps don't need these attributes.
+    delete this.appOptions.webapp.manifest;
+    delete this.appOptions.webapp.metaData;
+
+    utils.ensureFolderExists(utils.getFile(this.appOptions.STAGE_APP_DIR));
+
+    this.buildConfig.addConfig('APP_OPTIONS', this.appOptions);
+
+    // We have to rewrite below two ENV since they're using abspath and CURDIR.
+    this.buildConfig.addConfig('GAIA_DIR',
+      options.GAIA_DIR);
+    this.buildConfig.addConfig('XPCSHELLSDK',
+      options.XPCSHELLSDK);
+
+    // Create app's own makefile and store in STAGE_APP_DIR/app.mk.
+    this.appMkPath = utils.joinPath(this.appOptions.STAGE_APP_DIR, 'Makefile');
+    this.mainMake = new DependencyGraph(this.appMkPath);
+
+    this.genAppConfig();
+
+    this.mainMake.genBackend(this.buildConfig.getOutput('makefile'));
+  },
+
+  get execMakeCommand() {
+    return '@make -C ' + this.appOptions.STAGE_APP_DIR;
+  },
+
+  /**
+   * In order to run buildscript under each app, we need to use below command.
+   * XXX: we should try to fix it in xpcshell-commonjs.js.
+   *
+   * @param {string} moduleName - the module we would like to execute.
+   */
+  templateCommand: function(moduleName) {
+    return '@$(XPCSHELLSDK) -f $(GAIA_DIR)/build/xpcshell-commonjs.js ' +
+    '-e "run(\'' + moduleName + '\',$(APP_OPTIONS));"';
+  },
+
+  /**
+   * Generate app's own config data for build back-end.
+   */
+  genAppConfig: function() {
+    var buildFile = utils.getFile(this.appDir, 'build', 'build.js');
+    if (buildFile.exists()) {
+      // execute app's build.js if it has.
+      this.mainMake.insertTask(
+        'phony',
+        this.webapp.sourceDirectoryName + '-build',
+        this._preSteps,
+        ['@echo building ' + this.webapp.sourceDirectoryName + '\n\t' +
+        this.templateCommand(this.webapp.sourceDirectoryName + '/build')]
+      );
+    } else {
+      // if app has no build.js, we only copy the folder to build_stage.
+      this.mainMake.insertTask(
+        'phony',
+        this.webapp.sourceDirectoryName + '-build',
+        this._preSteps,
```

Comment:
If these two `insertTask` actually shared the same 3 arguments, we could only change the last argument by conditional expression or if...else statements.